FIO updates for supporting STM32MP1 I2C communication with SE050

mike-scott opened the pull request:
Patchset to do the following:
- Add support for STM32MP15 I2C5 in OP-TEE for using the Arduino headers on DK2
- Add support for STM32 I2C glue layer in SE050 crypto driver
- HACK to work around NACK handling issue in STM32 I2C driver

mike-scott commented:
Ugh, I just realized I made up a new DTS node where one already existed (near identical) in the kernel.

```
sed 's/se05x/se050/g'
```

I'll probably need to change that to the current standard.

mike-scott commented:
Diff from previous patchset:
```
diff --git a/core/drivers/crypto/se050/glue/i2c_stm32.c b/core/drivers/crypto/se050/glue/i2c_stm32.c
index 0482d772f..43482ab83 100644
--- a/core/drivers/crypto/se050/glue/i2c_stm32.c
+++ b/core/drivers/crypto/se050/glue/i2c_stm32.c
@@ -26,7 +26,7 @@ static int dt_get_se05x_node(void *fdt)
        static int node = -FDT_ERR_BADOFFSET;

        if (node == -FDT_ERR_BADOFFSET)
-               node = fdt_node_offset_by_compatible(fdt, -1, "nxp,se05x");
+               node = fdt_node_offset_by_compatible(fdt, -1, "nxp,se050");

        return node;
 }
```

ricardosalveti commented:
I think se05x might actually be better, because this same logic can be used with se051.

ricardosalveti commented on core/drivers/stm32_i2c.c:
```diff
@@ -1128,6 +1128,8 @@ static int i2c_write(struct i2c_handle_s *hi2c, struct i2c_request *request,
 	}
 
 	do {
+		udelay(25);
```
Just add a comment saying this is a temporary hack above this line.

mike-scott commented:
> I think se05x might actually be better, because this same logic can be used with se051.

Means we have different DTS than the kernel for the same devices tho.  Hard to say which direction on this.

mike-scott commented:
Looks like udelay(25) was a bit too aggressive.  Has to bump to udelay(50) to make other HW work.

mike-scott commented:
Diff from previous patchset:
```
diff --git a/core/drivers/stm32_i2c.c b/core/drivers/stm32_i2c.c
index 3be636515..a609032d1 100644
--- a/core/drivers/stm32_i2c.c
+++ b/core/drivers/stm32_i2c.c
@@ -1128,7 +1128,7 @@ static int i2c_write(struct i2c_handle_s *hi2c, struct i2c_request *request,
        }

        do {
-               udelay(25);
+               udelay(100);

                if (i2c_wait_txis(hi2c, timeout_ref))
                        goto bail;
```

mike-scott commented:
Added comment per @ricardosalveti :
```
diff --git a/core/drivers/stm32_i2c.c b/core/drivers/stm32_i2c.c
index a609032d1..eae075a46 100644
--- a/core/drivers/stm32_i2c.c
+++ b/core/drivers/stm32_i2c.c
@@ -1128,6 +1128,14 @@ static int i2c_write(struct i2c_handle_s *hi2c, struct i2c_request *request,
        }

        do {
+               /*
+                * This is a nasty timing hack.
+                * For an unknown reason there is a race condition when
+                * checking i2c_ack_failed from the i2c_write send loop.
+                * If a small udelay is added to the loop, this condition
+                * is avoided.
+                * Needs more research.
+                */
                udelay(100);

                if (i2c_wait_txis(hi2c, timeout_ref))
```

MrCry0 reviewed (approved):
LGTM

igoropaniuk commented:
I'd also suggest to add dynamic detection of bus id, which will be provided for i2c trampoline in Linux/U-Boot in RPC call, instead of `CFG_CORE_SE05X_I2C_BUS` provided during compile time. Something like this:

```
--- a/core/drivers/crypto/se050/glue/i2c.c
+++ b/core/drivers/crypto/se050/glue/i2c.c
@@ -17,7 +17,6 @@ static TEE_Result (*transfer)(struct rpc_i2c_request *req, size_t *bytes);
 static int i2c_transfer(uint8_t *buffer, int len, enum rpc_i2c_mode mode)
 {
        struct rpc_i2c_request request = {
-               .bus = CFG_CORE_SE05X_I2C_BUS,
                .chip = SMCOM_I2C_ADDRESS >> 1,
                .mode = mode,
                .buffer = buffer,
@@ -27,6 +26,7 @@ static int i2c_transfer(uint8_t *buffer, int len, enum rpc_i2c_mode mode)
        size_t bytes = 0;
        int retry = 5;
 
+       request.bus = native_get_i2c_bus();
        do {
                if ((*transfer)(&request, &bytes) == TEE_SUCCESS)
                        return bytes;
@@ -55,6 +55,11 @@ int glue_i2c_init(void)
        return native_i2c_init();
 }
 
+_weak int native_get_i2c_bus()
+{
+    return CFG_CORE_SE05X_I2C_BUS;
+}
+
 static TEE_Result load_trampoline(void)
 {
        if (IS_ENABLED(CFG_CORE_SE05X_I2C_TRAMPOLINE))

```

And in `core/drivers/crypto/se050/glue/i2c_stm32.c` provide `native_get_i2c_bus()` implementation, where bus id will be obtained from DTB.

Besides this LGTM.

ldts reviewed (changes requested):
I'd much rather avoid having to create a device tree node for the SE05x:.
I'd propose something like this instead:

```

The on the DTS

```
+++ b/core/arch/arm/dts/stm32mp157c-epc-revc.dts
@@ -23,6 +23,7 @@
     aliases {
         serial1 = &usart3;
         ethernet0 = &ethernet0;
+        i2c6 = &i2c6;
     };

``` 

#include <drivers/stm32_i2c.h>
#include <i2c_native.h>
#include <kernel/dt.h>
#include <kernel/boot.h>
#include <libfdt.h>
#include <phNxpEsePal_i2c.h>

static struct i2c_handle_s i2c;

static int dt_i2c_bus_config(struct stm32_i2c_init_s *init,
			     struct stm32_pinctrl **pctrl,  size_t *pcnt)
{
	const fdt32_t *cuint = NULL;
	const char *path = NULL;
	char bus[6] = { '\0' };
	void *fdt = NULL;
	int node = 0;

	fdt = get_embedded_dt();
	if (!fdt)
		return -FDT_ERR_NOTFOUND;

	snprintf(bus, sizeof(bus), "i2c%d", CFG_CORE_SE05X_I2C_BUS);

	path = fdt_get_alias(fdt, bus);
	if (!path) {
		EMSG("%d", __LINE__);
		return -FDT_ERR_NOTFOUND;
	}

	node = fdt_path_offset(fdt, path);
	if (node < 0) {
		EMSG("%d", __LINE__);
		return -FDT_ERR_NOTFOUND;
	}

	cuint = fdt_getprop(fdt, node, "clock-frequency", NULL);
	if (cuint && fdt32_to_cpu(*cuint) != CFG_CORE_SE05X_BAUDRATE)
		IMSG("SE05X ignoring CFG_CORE_SE05X_BAUDRATE, use DTB");
	else if (I2C_STANDARD_RATE != CFG_CORE_SE05X_BAUDRATE)
		IMSG("SE05x ignoring CFG_CORE_SE05X_BAUDRATE, use built-in");

	return stm32_i2c_get_setup_from_fdt(fdt, node, init, pctrl, pcnt);
}

int native_i2c_init(void)
{
	struct stm32_i2c_init_s i2c_init = { };
	struct stm32_pinctrl *pinctrl = NULL;
	size_t pin_count = 0;

	/* No need to re-init */
	if (i2c.i2c_state != I2C_STATE_RESET)
		return 0;

	/* support only one device on the platform */
	if (dt_i2c_bus_config(&i2c_init, &pinctrl, &pin_count))
		return -1;

	/* probe the device */
	i2c_init.own_address1 = SMCOM_I2C_ADDRESS;
	i2c_init.digital_filter_coef = 0;
	i2c_init.analog_filter = true;
	i2c.pinctrl_count = pin_count;
	i2c.pinctrl = pinctrl;

	stm32_i2c_resume(&i2c);

	return stm32_i2c_init(&i2c, &i2c_init);
}

TEE_Result native_i2c_transfer(struct rpc_i2c_request *req, size_t *bytes)
{
	if (req->mode == RPC_I2C_MODE_READ) {
		if (stm32_i2c_master_receive(&i2c, req->chip << 1, req->buffer,
					     req->buffer_len, 100))
			return TEE_ERROR_GENERIC;
	} else {
		if (stm32_i2c_master_transmit(&i2c, req->chip << 1, req->buffer,
					      req->buffer_len, 100))
			return TEE_ERROR_GENERIC;
	}

	*bytes = req->buffer_len;

	return TEE_SUCCESS;
}
```

ldts reviewed (changes requested):
I'd replace the udelay with this code.
The SE05x seems to oscilate between NACKF and TXIS...rather than hoping that 100u is enough, we should wait for TXIS using timeout_ref.

```
diff --git a/core/drivers/stm32_i2c.c b/core/drivers/stm32_i2c.c
index a609032d1..a5c1d2a80 100644
--- a/core/drivers/stm32_i2c.c
+++ b/core/drivers/stm32_i2c.c
@@ -925,8 +925,6 @@ static int i2c_wait_txis(struct i2c_handle_s *hi2c, uint64_t timeout_ref)
        while (!timeout_elapsed(timeout_ref)) {
                if (io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_TXIS)
                        break;
-           if (i2c_ack_failed(hi2c, timeout_ref))
-                   return -1;
        }
 
        if (io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_TXIS)
@@ -1128,8 +1126,6 @@ static int i2c_write(struct i2c_handle_s *hi2c, struct i2c_request *request,
        }
 
        do {
-           udelay(100);
-
                if (i2c_wait_txis(hi2c, timeout_ref))
                        goto bail;

```

mike-scott commented:
Changes to the patchset:
- Adjusted the hack patch and the glue layer per @ldts 
- Added a patch from @ldts which avoids re-initializing the glue layer
- Added a patch from @ldts which allows the trampoline to be disabled

Diff from the previous:
```
diff --git a/core/drivers/crypto/se050/glue/i2c.c b/core/drivers/crypto/se050/glue/i2c.c
index 20f838b2f..54243b733 100644
--- a/core/drivers/crypto/se050/glue/i2c.c
+++ b/core/drivers/crypto/se050/glue/i2c.c
@@ -54,6 +54,7 @@ int glue_i2c_init(void)
 	return native_i2c_init();
 }
 
+#ifndef CFG_NXP_SE05X_DISABLE_TRAMPOLINE
 static TEE_Result load_trampoline(void)
 {
 	transfer = rpc_io_i2c_transfer;
@@ -62,3 +63,4 @@ static TEE_Result load_trampoline(void)
 }
 
 boot_final(load_trampoline);
+#endif
diff --git a/core/drivers/crypto/se050/glue/i2c_stm32.c b/core/drivers/crypto/se050/glue/i2c_stm32.c
index 43482ab83..b565784ae 100644
--- a/core/drivers/crypto/se050/glue/i2c_stm32.c
+++ b/core/drivers/crypto/se050/glue/i2c_stm32.c
@@ -20,6 +20,7 @@
 /* Expect a single SE05x instance */
 static struct i2c_handle_s i2c_handle;
 static uint32_t se05x_i2c_addr;
+static uint8_t initialized;
 
 static int dt_get_se05x_node(void *fdt)
 {
@@ -88,6 +89,9 @@ int native_i2c_init(void)
 	size_t pin_count = 0;
 	struct stm32_i2c_init_s i2c_init = { };
 
+	if (initialized)
+		return 0;
+
 	ret = dt_se05x_i2c_config(&i2c_info, &pinctrl, &pin_count, &i2c_init);
 	if (ret < 0) {
 		EMSG("I2C configuration failed %d", ret);
@@ -112,8 +116,6 @@ int native_i2c_init(void)
 		panic();
 	}
 
-	stm32_i2c_resume(&i2c_handle);
-
 	if (!stm32_i2c_is_device_ready(i2c, se05x_i2c_addr,
 				       SE05X_I2C_TRIALS,
 				       SE05x_I2C_TIMEOUT_BUSY_MS)) {
@@ -121,6 +123,7 @@ int native_i2c_init(void)
 		panic();
 	}
 
+	initialized = 1;
 	return 0;
 }
 
diff --git a/core/drivers/stm32_i2c.c b/core/drivers/stm32_i2c.c
index eae075a46..d03c52696 100644
--- a/core/drivers/stm32_i2c.c
+++ b/core/drivers/stm32_i2c.c
@@ -922,6 +922,12 @@ static int i2c_ack_failed(struct i2c_handle_s *hi2c, uint64_t timeout_ref)
 /* Wait TXIS bit is 1 in I2C_ISR register */
 static int i2c_wait_txis(struct i2c_handle_s *hi2c, uint64_t timeout_ref)
 {
+#ifdef CFG_STM32_I2C_AVOID_EARLY_NACKF
+	while (!timeout_elapsed(timeout_ref)) {
+		if (io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_TXIS)
+			return 0;
+	}
+#else
 	while (!timeout_elapsed(timeout_ref)) {
 		if (io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_TXIS)
 			break;
@@ -931,6 +937,7 @@ static int i2c_wait_txis(struct i2c_handle_s *hi2c, uint64_t timeout_ref)
 
 	if (io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_TXIS)
 		return 0;
+#endif
 
 	if (i2c_ack_failed(hi2c, timeout_ref))
 		return -1;
@@ -1128,16 +1135,6 @@ static int i2c_write(struct i2c_handle_s *hi2c, struct i2c_request *request,
 	}
 
 	do {
-		/*
-		 * This is a nasty timing hack.
-		 * For an unknown reason there is a race condition when
-		 * checking i2c_ack_failed from the i2c_write send loop.
-		 * If a small udelay is added to the loop, this condition
-		 * is avoided.
-		 * Needs more research.
-		 */
-		udelay(100);
-
 		if (i2c_wait_txis(hi2c, timeout_ref))
 			goto bail;
 
```

mike-scott commented:
@ldts note I adjusted your new patch for the HACK to use `CFG_STM32_I2C_AVOID_EARLY_NACKF`

mike-scott commented on core/drivers/crypto/se050/glue/i2c_stm32.c:
```diff
@@ -0,0 +1,146 @@
+		panic();
+	}
+
+	stm32_i2c_resume(&i2c_handle);
```
done

mike-scott commented:
@ldts @igoropaniuk perhaps we can setup the DTS / parsing changes as a new PR on top of this one?

ricardosalveti commented:
@mike-scott I think this pr is not needed anymore with 3.18, right?
